compiler-plugin/src/main/java/io/ballerina/stdlib/xmldata/compiler/XmldataRecordFieldValidator.java

```diff
@@ -0,0 +1,151 @@
+    private static final String FLOAT = "float";
+    private static final String BOOLEAN = "boolean";
+    private static final String INT = "int";
+    private static final String QUESTION_MARK = "?";
+    private static final String TO_RECORD = "xmldata:toRecord";
+    private static final String VERTICAL_BAR = "|";
+    private static final String SQUARE_BRACKET = "[]";
+    private static final String BRACKET = "[";
+
+    @Override
+    public void perform(SyntaxNodeAnalysisContext ctx) {
+        List<Diagnostic> diagnostics = ctx.semanticModel().diagnostics();
+        for (Diagnostic diagnostic : diagnostics) {
+            if (diagnostic.diagnosticInfo().severity() == DiagnosticSeverity.ERROR) {
+                return;
+            }
+        }
+        Node node = ctx.node();
+        if (node instanceof RecordFieldNode) {
+            this.recordNodes.add((RecordFieldNode) node);
+        }
+        if (node instanceof VariableDeclarationNode) {
+            VariableDeclarationNode variableDeclarationNode = (VariableDeclarationNode) node;
+            Optional<ExpressionNode> initializer = variableDeclarationNode.initializer();
+            if (!initializer.isEmpty()) {
+                ExpressionNode expressionNode = initializer.get();
+                if (expressionNode.toString().contains(TO_RECORD)) {
+                    TypedBindingPatternNode typedBindingPatternNode = variableDeclarationNode.typedBindingPattern();
+                    checkRecordField(typedBindingPatternNode.typeDescriptor().toString(), ctx);
+                }
+            }
+        }
+        if (node instanceof ModuleVariableDeclarationNode) {
+            ModuleVariableDeclarationNode moduleVariableDeclarationNode = (ModuleVariableDeclarationNode) node;
+            Optional<ExpressionNode> initializer = moduleVariableDeclarationNode.initializer();
+            if (!initializer.isEmpty()) {
+                ExpressionNode expressionNode = initializer.get();
+                if (expressionNode.toString().contains(TO_RECORD)) {
+                    TypedBindingPatternNode typedBindingPatternNode =
+                            moduleVariableDeclarationNode.typedBindingPattern();
+                    checkRecordField(typedBindingPatternNode.typeDescriptor().toString(), ctx);
+                }
+            }
+        }
+    }
+
+    private void checkRecordField(String recordName, SyntaxNodeAnalysisContext ctx) {
+        for (RecordFieldNode recordFieldNode: this.recordNodes) {
+            if (!this.validatedRecordNodes.contains(recordFieldNode)) {
+                String recordNameOfField = recordFieldNode.parent().parent().children().get(1).toString().trim();
+                String filedType = recordFieldNode.typeName().toString().trim();
+                if (recordNameOfField.equals(recordName.trim())) {
+                    this.validatedRecordNodes.add(recordFieldNode);
+                    if (filedType.contains(QUESTION_MARK)) {
```

Comment:
Are we checking if it is an optional type?
If so, I think the proper way is to check it through the syntax tree. eg:
```
if (recordFieldNode.questionMarkToken() != null) {
    // logic goes here
}
```

Let's try to avoid string searches as much as possible. Check other places as well.

Reply:
This is used to check whether the field is optional or not. Here, we have to check whether the field type is optional or not.
So, I tried to remove `toString` usage by checking the kind of node but it is not worked correctly.
- If the union type has an optional type, we can't check the union type by using kind().
- If the type is a union type and that hasn't any optional field, we can divide the type by using `rightTypeDesc()` or `leftTypeDesc()` but if the type contains multiple types, it is hard.
Therefore, I haven't updated this impl.

Reply:
Ack.

Reply:
@kalaiyarasiganeshalingam what happen if user define the record type like below,

```ballerina
type Test record {
    Bar|Bar1|() foo;
};
```

Reply:
with a string value check, there is a possibility to miss certain scenarios.

Reply:
Fixed in [cf5bacd](https://github.com/ballerina-platform/module-ballerina-xmldata/pull/430/commits/cf5bacdb016ef7ecc11ecfa7d9866e832e9bbbb0)